src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -128,15 +128,17 @@ public static async ValueTask<ImmutableArray<(Checksum, object)>> ReadDataAsync(
             catch (EndOfStreamException)
             {
                 cancellationToken.ThrowIfCancellationRequested();
-
-                throw exception ?? ExceptionUtilities.Unreachable;
             }
             finally
             {
                 // Make sure to complete the copy and pipes before returning, otherwise the caller could complete the
                 // reader and/or writer while they are still in use.
                 await copyTask.ConfigureAwait(false);
             }
+
+            // The stream ended unexpectedly before we read all the data from the local pipe.
+            // copyTask should have completed with an exception and that should have thrown when we awaited the task.
+            throw ExceptionUtilities.Unreachable;
```

Comment:
not certain i get this either tbh... but this i feel less clear on.

What really worries me is just how much cmplexity there is around using a Pipe (which i thought was supposed to be a fairly simple abstraction to make it easier to do things right).  I'm strongly wondering if we should pull in some core BCL people (maybe Stephen) to discuss the matter and get some audits going on if we're doing things appropriately here, or if there's something we're missing to get this right.

--

Put another way.  There's no way i would have detected a problem with the 'old' code.  And i don't have any way of really validating to myself that hte new code is 'correct' (or even 'more correct' than before).  The situation feels very pit-of-fail-y to me, and i think the only way to dig ourselves out of that is either to get an expert to walk us through things, or for someone on the team to figure everything out in minute detail, and then determinate the information to the rest of the team.'

Note: this is what we did with async/Tasks when we starting using them.  We had similar issues around our intuitions and expectations not being held.  But we were able to eventually get a full understanding (including pushing tweaks down into the runtime) so we could effectively use those APIs correctly.

Reply:
I agree. I don't quite understand all the interactions either :(

Reply:
@AArnott FYI

Reply:
Ideally, we could get a snippet showing how to read/write from a pipe in a rock-solid fashion.  Including being absolutely bullet-proof for all the sorts of failures that can happen (esp. cancellation).  

Thanks!

Reply:
@AArnott Is familiar with this use case - @sharwell, @AArnott and I have spend hours looking into this code multiple times. 

The constraint we have that I believe we should remove is that our solution asset serializer reads/writes synchronously. That prevents us from reading/writing to the RPC pipe directly and instead we need extra local pipe on each end to mediate the reads/writes. If we rewrote our serializer to be async this could would significantly simplify. Unfortunately that's quite a bit of work and I have not had the time to do it yet. I'd certainly like to.